cached_addr_book.go

```diff
@@ -0,0 +1,307 @@
+package main
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/ipfs/boxo/routing/http/types"
+	"github.com/libp2p/go-libp2p-kad-dht/amino"
+	"github.com/libp2p/go-libp2p/core/event"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/p2p/host/peerstore/pstoremem"
+	"github.com/libp2p/go-libp2p/p2p/protocol/circuitv2/relay"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	Subsystem = "cached_addr_book"
+	// The TTL to keep recently connected peers for. Same as [amino.DefaultProvideValidity] in go-libp2p-kad-dht
+	RecentlyConnectedAddrTTL = amino.DefaultProvideValidity
+
+	// Connected peers don't expire until they disconnect
+	ConnectedAddrTTL = peerstore.ConnectedAddrTTL
+
+	// How long to wait since last connection before probing a peer again
+	PeerProbeThreshold = time.Hour
+
+	// How often to run the probe peers loop
+	ProbeInterval = time.Minute * 15
+
+	// How many concurrent probes to run at once
+	MaxConcurrentProbes = 20
+
+	// How long to wait for a connect in a probe to complete.
+	// The worst case is a peer behind a relay, so we use the relay connect timeout.
+	ConnectTimeout = relay.ConnectTimeout
+
+	// How many peers to cache in the peer state cache
+	// 100_000 is also the default number of signed peer records cached by the memory address book.
+	PeerCacheSize = 100_000
```

Comment:
You can make this much larger(1M?). Compared to the peerstore the info stored here is much smaller. Also this has negative results for failed queries.